server/resp/src/main/java/org/infinispan/server/resp/serialization/lua/LuaResponseWriter.java

```diff
@@ -0,0 +1,285 @@
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+
+import org.infinispan.commons.jdkspecific.CallerId;
+import org.infinispan.server.resp.RespVersion;
+import org.infinispan.server.resp.serialization.JavaObjectSerializer;
+import org.infinispan.server.resp.serialization.Resp3Type;
+import org.infinispan.server.resp.serialization.ResponseWriter;
+import org.infinispan.server.resp.serialization.SerializationHint;
+
+import party.iroiro.luajava.Lua;
+
+public class LuaResponseWriter implements ResponseWriter {
+   private final Lua lua;
+   private RespVersion version = RespVersion.RESP3; // TODO: once we support RESP2, this should be lowered
+
+   public LuaResponseWriter(Lua lua) {
+      this.lua = lua;
+   }
+
+   @Override
+   public RespVersion version() {
+      return version;
+   }
+
+   @Override
+   public void version(RespVersion version) {
+      this.version = version;
+   }
+
+   @Override
+   public boolean isInternal() {
+      return true;
+   }
+
+   @Override
+   public void nulls() {
+      lua.checkStack(1);
+      lua.pushNil();
+   }
+
+   @Override
+   public void ok() {
+      lua.checkStack(1);
+      lua.push("OK");
+   }
+
+   @Override
+   public void queued(Object ignore) {
+      throw new UnsupportedOperationException("queued(Object)");
+   }
+
+   @Override
+   public void simpleString(CharSequence value) {
+      lua.checkStack(1);
+      if (value == null) {
+         lua.pushNil();
+      } else {
+         lua.push(value.toString());
+      }
+   }
+
+   @Override
+   public void string(CharSequence value) {
+      lua.checkStack(1);
+      if (value == null) {
+         lua.pushNil();
+      } else {
+         lua.push(value.toString());
+      }
+   }
+
+   @Override
+   public void string(byte[] value) {
+      lua.checkStack(1);
+      if (value == null) {
+         lua.pushNil();
+      } else {
+         lua.push(new String(value, StandardCharsets.US_ASCII));
+      }
+   }
+
+   @Override
+   public void integers(Number value) {
+      lua.checkStack(1);
+      if (value == null) {
+         lua.pushNil();
+      } else {
+         lua.push(value);
+      }
+   }
+
+   @Override
+   public void doubles(Number value) {
+      lua.checkStack(1);
+      if (value == null) {
+         lua.pushNil();
+      } else {
+      lua.push(value);
+      }
+   }
+
+   @Override
+   public void booleans(boolean value) {
+      lua.checkStack(1);
+      lua.push(value);
+   }
+
+   @Override
+   public void arrayEmpty() {
+      lua.checkStack(1);
+      lua.push(false);
+   }
+
+   @Override
+   public <T> void array(Collection<T> collection, JavaObjectSerializer<T> serializer) {
+      if (collection == null) {
+         nulls();
+      } else {
+         lua.checkStack(2);
+         lua.newTable();
+         int i = 1;
+         for (T o : collection) {
+            lua.push(i);
+            serializer.accept(o, this);
+            lua.setTable(-3);
```

Comment:
Have to admit all these magical numbers are not helpful. Maybe we should have these special values on an interface commenting what they mean?

Reply:
nothing magical here: https://www.lua.org/manual/5.1/manual.html#lua_settable

Reply:
It is magical in the sense that we have to look it up somewhere else that isn't in the code. I know it has a defined purpose.

Reply:
I have sprinkled a bunch of comments which should make things clearer